src/bcd.c

```diff
@@ -581,6 +602,91 @@ bcd_kv_set(struct bcd_session *session, struct bcd_packet *packet)
 	    "malformed key-value pair");
 }
 
+static ssize_t
+bcd_arg_get(char **output, size_t n_output, bcd_error_t *error)
+{
+	struct bcd_arg *cursor;
+	size_t limit = n_output;
+	size_t i = 0;
+
+	if (bcd_arg_count == 0) {
+		return 0;
+	}
+
+	if (n_output > bcd_arg_count)
+		limit = bcd_arg_count;
+
+	if (limit > BCD_ARGC_LIMIT)
+		limit = BCD_ARGC_LIMIT;
+
+	TAILQ_FOREACH(cursor, &bcd_arg_list, linkage) {
+		int ra;
+
+		if (i == limit)
+			break;
+
+		ra = asprintf(&output[i++], "%s", cursor->arg);
+		if (ra == -1) {
+			bcd_error_set(error, 0, "failed to allocate arg");
+			goto fail;
+		}
+	}
+
+	return (ssize_t)i;
+fail:
+	while (i-- > 0)
+		free(output[i]);
+
+	return -1;
+}
+
+static int
+bcd_arg_set(struct bcd_session *session, struct bcd_packet *packet)
+{
+	const char *arg = packet->payload;
+	struct bcd_arg *argp, *cursor;
+	size_t arglen;
+	char *stream;
+
+	arglen = strlen(arg);
+	if (arglen == 0)
+		goto fail;
+
+	argp = malloc(sizeof(*argp) + arglen + 1);
+	if (argp == NULL) {
+		return bcd_error(BCD_EVENT_METADATA, session,
+		    "internal memory allocation error");
+	}
+
+	if (bcd_arg_count == 0) {
+		TAILQ_INIT(&bcd_arg_list);
+	} else {
+		struct bcd_arg *tmp;
+
+		TAILQ_FOREACH_SAFE(cursor, &bcd_arg_list, linkage, tmp) {
```

Comment:
Why do you need `FOREACH_SAFE` here? You break after cleanup so as far as I can tell, TAILQ_FOREACH should be sufficient.